server/src/fstar_connection.ts

```diff
@@ -0,0 +1,512 @@
+		this.fstar.proc.stderr?.on('data', (data) => { console.error(`${fstar_proc_name} stderr: ${data}`); });
+	}
+
+	// Attempts to spawn an F* process, using the given configuration and
+	// filePath, and create a connection to it.
+	//
+	// @throws {Error} from `trySpawnFstar`
+	static tryCreateFStarConnection(fstarConfig: FStarConfig, filePath: URI, debug: boolean, lax?: 'lax') : FStarConnection | undefined {
+		const fstar = FStar.trySpawnFstar(fstarConfig, filePath, debug, lax);
+		if (fstar)
+			return new FStarConnection(fstar);
+		else
+			return undefined;
+	}
+
+	// Kills the F* process and closes the connection
+	close() {
+		this.fstar.proc.kill();
+	}
+
+	async restartSolver(debug: boolean) {
+		this.fstar.killZ3SubProcess(debug);
+
+		// Wait for a second for processes to die before restarting the solver
+		await setTimeout(1000);
+		this.restartSolverRequest(debug);
+	}
+
+	// Gets the configuration for the F* process.
+	fstar_config() : FStarConfig {
+		return this.fstar.config;
+	}
+
+	////////////////////////////////////////////////////////////////////////////////////
+	// Methods to send requests to F*
+	////////////////////////////////////////////////////////////////////////////////////
+
+	// Sends a request to the F* process.
+	//
+	// @param expect_response indicates if a response from F* is expected for
+	// this request.
+	private async request(msg: any, debug: boolean, expect_response = true, is_stream = false): Promise<any> {
+		const qid = this.last_query_id + 1;
+		this.last_query_id = qid;
+		msg["query-id"] = '' + qid;
+		const text = JSON.stringify(msg);
+		if (debug) {
+			console.log(">>> " + text);
+		}
+		if (this.fstar.proc.exitCode != null) {
+			const process_name = this.fstar.lax ? "flycheck" : "checker";
+			const error_msg = "ERROR: F* " + process_name + " process exited with code " + this.fstar.proc.exitCode;
+			// TODO(klinvill): should we throw exceptions here or reject the
+			// returned promise? Although I believe the semantics of async
+			// functions are such that this will automatically result in a
+			// rejected promise?
+			throw new Error(error_msg);
+		}
+
+		try {
+			this.fstar.proc?.stdin?.write(text);
+			this.fstar.proc?.stdin?.write("\n");
+		} catch (e) {
+			// TODO(klinvill): should we throw exceptions here or reject the
+			// returned promise? Although I believe the semantics of async
+			// functions are such that this will automatically result in a
+			// rejected promise?
+			const msg = "ERROR: Error writing to F* process: " + e;
+			throw new Error(msg);
+		}
+
+		const internal_promise =  withResolvers();
+		if (expect_response) {
+			// Only keep track of responses if we expect a response
+			this.pending_responses.set(qid, {ip: internal_promise, is_stream});
+		}
+		// TODO(klinvill): Should we always return a promise even if no response
+		// is expected? In that case the promise is just ignored.
+		return internal_promise.promise;
```

Comment:
I think it would be shorter and cleaner to write this without the additional wrappers:
```suggestion
		if (expect_response) {
		        return new Promise((resolve, reject) =>
		                this.pending_responses.set(qid, {resolve, reject, is_stream}));               
		}
```

> Should we always return a promise even if no response is expected?

If we stored it in `this.pending_responses`, then that would lead to a memory leak (since we'll never receive a response that would delete it).

Reply:
> Should we always return a promise even if no response is expected?

Sorry, I think I misunderstood the question here.

I believe what you're curious about is the following:
```js
async sendNotification() {
  return new Promise((resolve, reject) => {
    /* discard the resolve handler because we don't expect a response */
  });
}
```

This is not a good idea.  When you call `await sendNotification()`, the caller will then wait forever because the promise is not resolved.